HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.h

```diff
@@ -45,6 +46,8 @@ class HLTMuonDimuonL3Filter : public HLTFilter {
                             const reco::RecoChargedCandidateRef&,
                             const reco::BeamSpot&,
                             const edm::ESHandle<MagneticField>&) const;
+
+  mutable PropagateToMuon prop_;
```

Comment:
I wonder if the `mutable` attribute here might clash with the need to use this object inside a `global` module (which `HLTFilter` is). Maybe one step in the right direction is to use `std::atomic`. I can't judge with certainty, so I would kindly ask advise from @cms-sw/core-l2.

Reply:
On a quick look `PropagateToMuon::init()` appears to set various member variables, and intended to be called on each event, so the use of `mutable` here indeed looks unsafe. The `std::atomic` won't help for many reasons, and any synchronization mechanism (like `std::mutex`) would be bad because the critical section is essentially the duration of the entire `filter()` method.

Here are some alternatives (in no particular order)
* place `PropagateToMuon` object into an `edm::StreamCache` (or `edm::stream` module type)
* refactor `PropagateToMuon` into two classes such that
  * one class contains the current constructor and the `init()` function (as `const`), but instead of writing to member variables, it would create an object of an another class (below)
  * another class that contains the member variables currently being written to in `init()`, and all the `extrapolate() const` methods
* move `PropagateToMuon` into EventSetup, in which case the `ESProducer` would take the role of the current constructor, and the `ESProcucer::produce()` + the constructor of `PropagateToMuon` would take the role of the current `init()` method

Reply:
Thanks, @makortel . @JanFSchulte , I think this point will need to be addressed before moving forward with this PR.

Reply:
Thanks for the feedback @makortel @missirol . Given that the impact on the HLT performance is minimal and we included this update mostly because it is technically the more correct way to perform the matching, I am not sure the effort is worth the gain. We might end up going with the just the fix to the L1 filter that is unaffected by this question and is an actual performance improvement.